### PR TITLE
Codereorg: Part 5, state 

### LINKED ISSUE #1461: Strange shutdown/restart behavior on 0.22.2
## Bug description

After upgrading to 0.22.2 restart or shutdown take like over 10 minutes on latest Debian. Installed from *.deb. Working like a charm on 0.22.1.

## Environment

- OS: Debian
- Headscale version: 0.22.2
- Tailscale version: 1.40.1 (not relevant)

- [ ] Headscale is behind a (reverse) proxy
- [ ] Headscale runs in a container

## To Reproduce

Upgrade, do `systemd restart headscale`, wait like for 10 minutes to restart.

### PR DESCRIPTION
**Pull request by kradalby**
This PR reworks the state management in headscale, instead of relying on timestamps and frequent checking, and sending the entire world every time, we now send incremental updates based on the change that has happened. 

Closes #1390
Fixes #1461

### CODE REVIEW

**Comment by kradalby** [comment]
All tests pass locally, it is hard to know if Github actions is reliable or not.

**Comment by kradalby** [comment]
My observation so far is that tests including a larger amount of nodes are more unreliable, particularly with excess network requirements:
- PingAll (the base of the typically failed ones)
- Taildrop

File: hscontrol/mapper/mapper.go [hunk 1]
```diff
@@ -302,84 +293,118 @@ func (m Mapper) CreateMapResponse(
...
+	mapRequest tailcfg.MapRequest,
+	machine *types.Machine,
+	machineKeys []uint64,
+	pol *policy.ACLPolicy,
+) ([]byte, error) {
+	var err error
+	changed := make(types.Machines, len(machineKeys))
+	lastSeen := make(map[tailcfg.NodeID]bool)
+	for idx, machineKey := range machineKeys {
+		peer, err := m.db.GetMachineByID(machineKey)
```
Comment (juanfont) [inline hscontrol/mapper/mapper.go]: Mixing machinekey with machineIDs ?

**Comment by vsychov** [comment]
Hello there,
Thank you for your hard work on headscale!

I'm start testing this changes, and found bug with "on-the-fly" state update:
1. Start headscale.
2. Connect client (macos): `/Applications/Tailscale.app/Contents/MacOS/Tailscale up --login-server=https://headscale-test.example.com/`
3. Connect `headscale-test-1` (linux): `tailscale login --login-server=https://headscale-test.example.com/ --advertise-exit-node`
4. Connect `headscale-test-ams3-1` and `headscale-test-ams3-2` (linux): `tailscale up --login-server=https://headscale-test.example.com/ --advertise-routes=10.0.0.0/8 --ssh`
5. Approve the routes and verify:
```
ID | Machine               | Prefix     | Advertised | Enabled | Primary
1  | headscale-test-1      | 0.0.0.0/0  | true       | true    | -
2  | headscale-test-1      | ::/0       | true       | true    | -
4  | headscale-test-ams3-2 | 10.0.0.0/8 | true       | true    | true
3  | headscale-test-ams3-1 | 10.0.0.0/8 | true       | true    | false
```
6. Try to ping: 
```
/Applications/Tailscale.app/Contents/MacOS/Tailscale ping 10.110.0.53
2023/09/15 23:06:47 no matching peer
```

7. 
```
➜  ~ /Applications/Tailscale.app/Contents/MacOS/Tailscale status
100.64.0.1      node1   user.example.com macOS   -
100.64.0.2      headscale-test-1     user.example.com linux   -
100.64.0.3      headscale-test-ams3-1..example.com user.example.com linux   -
100.64.0.4      headscale-test-ams3-2..example.com user.example.com linux   -
```

8. The exit-node on `headscale-test-1` does not appear, and the ping to `10.110.0.53` does not work (though it should through `headscale-test-ams3-1`).
9. The problem is resolved by restarting headscale, but the status does not update until the restart (I wait ~10 min until restart headscale).

**Comment by vsychov** [comment]
It seems there is also an issue with updating the "last-seen"/"online" status. I wanted to test the subnet-router failover, but as soon as I executed `tailscale down` on `headscale-test-ams3-2`, I expected the traffic to switch through `headscale-test-ams3-1`. However, `headscale-test-ams3-1` was unexpectedly marked as "offline":
```
root@headscale-f7489674b-4j5jf:/# date
Fri Sep 15 21:30:43 UTC 2023
root@headscale-f7489674b-4j5jf:/# headscale nodes list
ID | Hostname              | Name                  | MachineKey | NodeKey | User                 | IP addresses                  | Ephemeral | Last seen           | Expiration          | Online  | Expired
....
3  | headscale-test-ams3-1 | headscale-test-ams3-1 | [K4v5r]    | [LkHjz] | user.example.com | 100.64.0.3, fd7a:115c:a1e0::3 | false     | 2023-09-15 21:14:13 | 2023-09-16 16:58:33 | offline | no
....
root@headscale-f7489674b-4j5jf:/# 
```

Although there are no network connectivity issues, tailscale is working, and headscale is accessible via http:
```
root@headscale-test-ams3-1:~# systemctl status tailscaled.service
● tailscaled.service - Tailscale node agent
     Loaded: loaded (/lib/systemd/system/tailscaled.service; enabled; vendor preset: enabled)
     Active: active (running) since Fri 2023-09-15 20:57:34 UTC; 32min ago
       Docs: https://tailscale.com/kb/
    Process: 8483 ExecStartPre=/usr/sbin/tailscaled --cleanup (code=exited, status=0/SUCCESS)
   Main PID: 8511 (tailscaled)
     Status: "Connected; user.example.com; 100.64.0.3 fd7a:115c:a1e0::3"
      Tasks: 8 (limit: 1116)
     Memory: 39.5M
        CPU: 1.760s
     CGroup: /system.slice/tailscaled.service
             └─8511 /usr/sbin/tailscaled --state=/var/lib/tailscale/tailscaled.state --socket=/run/tailscale/tailscaled.sock --port=41641
root@headscale-test-ams3-1:~# curl -I https://headscale-test.example.com/api/v1/machine
HTTP/2 401
date: Fri, 15 Sep 2023 21:30:33 GMT
content-type: text/plain; charset=utf-8
content-length: 12
strict-transport-security: max-age=15724800; includeSubDomains

root@headscale-test-ams3-1:~#
```

**Comment by kradalby** [comment]
@vsychov Thank you, I appreciate this a lot. I think what I will do since this PR is so messy already is to get it in, and then it would be great if you could create two issues for me on this (and other things you find), and then we make them blocking before alpha.

I would also appreciate your continued testing of this work.

**Comment by juanfont** [comment]
Approved!

**Review by juanfont — APPROVED** [review]

**Comment by kradalby** [comment]
@vsychov did you try the tests with the route/exit node issue including the last commit? Would you mind opening an issue for that one?

I'll have a look at the online status now.

**Comment by vsychov** [comment]
@kradalby , I'll try my tests on current `main` revision, and if that still reproducible, I'll create issues for that.

**Comment by vsychov** [comment]
Problems still appear, I created issue: #1561

**Pull request merged**